### ISSUE TITLE
RStudio Connections Pane reacts excessively slowly and freezes RStudio when connected to an Oracle database using `package:odbc`

### ISSUE THREAD

**Issue by ThomasSoeiro** [rstudio/rstudio#12699]
### System details

    RStudio Edition : RStudio Pro
    RStudio Version : 2022.12.0 Build 353.pro20 
    OS Version : Red Hat Enterprise Linux Server 7.8 (Maipo)
    R Version : 4.1.2 (2021-11-01)

### Steps to reproduce the problem

1. Connect to an Oracle database using `package:DBI`, `package:odbc`, and RStudio's Professional driver;
2. The Connections Pane shows all schemas rather than the schema of the connected user;
3. Exploring the content of the database through the Connections Pane is excessively long (e.g. expanding schemas);
4. RStudio become unresponsive in the meantime.

I don't know if this is related to RStudio, RStudio's Professional drivers, or `package:odbc`. I asked my IT department to [set up Oracle's OEM driver](https://solutions.posit.co/connections/db/databases/oracle/) to test it. I will report the results if they do. I have no idea how to investigate the issue further but would be happy to provide more information. 

### Describe the problem in detail

```r
library(DBI)
con <- dbConnect(odbc::odbc(), dsn = "xxx")
```

Once connected, the Connections Pane shows all schemas rather than the schema of the connected user. This has been an issue for other users too (e.g. [here](https://community.rstudio.com/t/rstudio-oracle-db-connection/129118) and [here](https://stackoverflow.com/a/65505785)).

Expanding schemas (e.g. the schema of the connected user) is excessively long and freezes RStudio. The issue may be related to r-dbi/odbc#540.

Commands issued in RStudio Console in the meantime (i.e. during the time that elapses between clicking on a schema and its expansion) are executed only after the schema has expanded.

Taken together, this makes the Connections Pane useless and unusable to show the content of the database.

[`package:connections`](https://github.com/rstudio/connections) is not installed on my system.

### Describe the behavior you expected

1. The Connections Pane should allow explore the content of the database [as per the documentation](https://solutions.posit.co/connections/db/tooling/connections/) without freezing RStudio.
2. When connected to an Oracle database, the Connections Pane should probably show the schema of the connected user by default.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ThomasSoeiro** [comment]
> I don't know if this is related to RStudio or RStudio's Professional drivers. I asked my IT department to [set up Oracle's OEM driver](https://solutions.posit.co/connections/db/databases/oracle/) to test it. I will report the results if they do. I have no idea how to investigate the issue further but would be happy to provide more information. 

After testing, there is another issue with Oracle's OEM driver. I don't know if it is related or not.  It is needed to wrap `dbConnect()` in e.g. `invisible()` or `system.time()`, otherwise the connection never establishes. In this case, the Connection Pane is not populated by the database content.

```r
library(DBI)

system.time(
  con <- dbConnect(drv = odbc::odbc(), dsn = "yyy")
)
# utilisateur     système      écoulé 
#       0.026       0.008       0.221 

start_time <- Sys.time()
con <- dbConnect(drv = odbc::odbc(), dsn = "yyy")
start_time - Sys.time()
# still running after 15+ minutes
```

**Comment by ronblum** [comment]
@ThomasSoeiro Thank you for the added information! For what it's worth, showing all of the Oracle accounts had always been the behavior for Oracle and a couple of other data sources, but most only show the user tables. I'm trying to find out where this is controlled. Note that for Oracle, even before the Connections Pane, `DBI::dbListTables(con)` will return all tables. 

Regarding running in R, outside of RStudio: If you open up the Terminal in RStudio Server, are you able to start R and connect from there?

**Comment by ThomasSoeiro** [comment]
> For what it's worth, showing all of the Oracle accounts had always been the behavior for Oracle and a couple of other data sources, but most only show the user tables. I'm trying to find out where this is controlled.

Although this is not the main issue, an option to control the default view would be nice to have.

> Note that for Oracle, even before the Connections Pane, `DBI::dbListTables(con)` will return all tables.

`dbListTables()` is currently unusable with `package:odbc` (see r-dbi/odbc#540). Since the Connection Pane is not either in a working state, working with an Oracle database with `package:DBI` and RStudio is quite difficult.

> Regarding running in R, outside of RStudio: If you open up the Terminal in RStudio Server, are you able to start R and connect from there?

I can't find a [terminal](https://www.rstudio.com/blog/rstudio-v1-1-preview-terminal/) in my version of RStudio. Maybe I'm not looking in the right place, or it has been disabled by my IT department?

![Capture](https://user-images.githubusercontent.com/124079058/220574772-1be16cc9-7d61-44a9-a385-0f2a5a6e1751.PNG)

**Comment by ThomasSoeiro** [comment]
> I can't find a [terminal](https://www.rstudio.com/blog/rstudio-v1-1-preview-terminal/) in my version of RStudio. Maybe I'm not looking in the right place, or it has been disabled by my IT department?

@ronblum Could you please either give further guidance to open a terminal or confirm if it is disabled in my version? In the latter case, I will ask my IT department to run the test.

Thank you!

**Comment by gtritchie** [comment]
The lack of a `Tools / Terminal` menu means your IT department has disabled this feature. That is done with the `rsession.conf` setting `allow-shell=0`. I'm not aware of any other way this menu would be missing.

**Comment by zedapolo** [comment]
Directly from command line we have the following result:

> library(DBI)
>
> start_time <- Sys.time()
> con <- dbConnect(drv = odbc::odbc(), dsn = "yyy")
> Sys.time() - start_time
Time difference of 20.55364 secs
> dbDisconnect(con)
>
> system.time(
+   con <- dbConnect(drv = odbc::odbc(), dsn = "yyy")
+ )
utilisateur     système      écoulé
      0.015       0.008       0.232

**Comment by ThomasSoeiro** [comment]
According to @detule in r-dbi/odbc#540 and r-dbi/odbc#560 the performance issue in RStudio Connections Pane is due to an issue in `package:odbc`.

In addition, there is also the issue that a connection to an Oracle database is excessively long to establish when using Oracle's OEM driver (i.e. `dbConnect(drv = odbc::odbc(), dsn = "<DSN using Oracle's OEM driver>")`, see https://github.com/rstudio/rstudio/issues/12699#issuecomment-1435754705 for more information).

**Comment by detule** [comment]
Hi -

Just a note that we rolled out an ORACLE performance enhancement for Oracle in `package:odbc 1.3.5` and closed the [corresponding issue](https://github.com/r-dbi/odbc/issues/540) there.

Thanks

**Comment by ThomasSoeiro** [comment]
I confirm that the connection pane now reacts as quick as expected with `package:odbc` 1.3.5 when using RStudio's Professional drivers.

However, the issue reported with Oracle's OEM driver persists (see https://github.com/rstudio/rstudio/issues/12699#issuecomment-1435754705 for more information).

**Comment by ThomasSoeiro** [comment]
The issue that prompted the report is now fixed in `package:odbc` 1.3.5 (see r-dbi/odbc#540). The issue can be closed.

I reported the two other issues that we discussed here in separate reports:

- [the issue regarding the Connections Pane usability](https://github.com/rstudio/rstudio/issues/12699#issuecomment-1439055253) in r-dbi/odbc#605;
- and [the issue regarding the connection using Oracle's OEM driver](https://github.com/rstudio/rstudio/issues/12699#issuecomment-1435754705) in r-dbi/odbc#606.

**Comment by ronblum** [comment]
@ThomasSoeiro Thank you for the update! Now to comment on the related issue. 🙂

**Comment by ronblum** [comment]
@ThomasSoeiro See also #13672.
